node/src/components/linear_chain.rs

```diff
@@ -234,36 +247,76 @@ where
                     public_key,
                     ..
                 } = *fs;
+                // TODO: Also verify that the public key belongs to a bonded validator!
                 if let Err(err) = fs.verify() {
-                    error!(%block_hash, %public_key, %err, "received invalid finality signature");
+                    warn!(%block_hash, %public_key, %err, "received invalid finality signature");
                     return Effects::new();
                 }
                 debug!(%block_hash, %public_key, "received new finality signature");
-                async move {
-                    let maybe_block = effect_builder
-                        .get_block_from_storage(block_hash)
-                        .await;
-                    match maybe_block {
-                        None => {
-                            let block_hash = fs.block_hash;
-                            let public_key = fs.public_key;
-                            warn!(%block_hash, %public_key, "received a signature for a block that was not found in storage");
-                        }
-                        Some(mut block) => {
-                            if !block.proofs().iter().any(|proof| proof == &fs.signature) {
-                                block.append_proof(fs.signature);
-                                let _ = effect_builder
-                                    .put_block_to_storage(Box::new(block))
-                                    .await;
-                                let _ = effect_builder
-                                    .broadcast_message(Message::FinalitySignature(fs.clone()))
-                                    .await;
-                                let _ = effect_builder.announce_finality_signature(fs).await;
-                            }
-                        }
-                    }
+                let sigs = self
+                    .pending_finality_signatures
+                    .entry(public_key)
+                    .or_default();
+                // Limit the memory we use for storing unknown signatures from each validator.
+                if sigs.len() >= MAX_PENDING_FINALITY_SIGNATURES_PER_VALIDATOR {
+                    warn!(
+                        %block_hash, %public_key,
+                        "received too many finality signatures for unknown blocks"
+                    );
+                    return Effects::new();
+                }
```

Comment:
Once we hit the limit for a validator we will never add another signature from it. That's b/c we only purge entries `pending_finality_signatures` collection after receiving `FinalitySignatureReceived` event. We should purge it also when we receive a new linear chain block.

Reply:
Ah, right, good catch! :man_facepalming:  :+1:

Reply:
Done in https://github.com/CasperLabs/casper-node/pull/649/commits/77663e4d9aa78c0b36a5d6be3ca54fa1860f48a1.